metrics/resource_view.go

```diff
@@ -0,0 +1,306 @@
+// Tags are.
+type ResourceExporterFactory func(*resource.Resource) (view.Exporter, error)
+type meters struct {
+	meters  map[string]*meterExporter
+	factory ResourceExporterFactory
+	// Cache of Resource pointers from metricskey to Meters, to avoid
+	// unnecessary stringify operations
+	resourceToKey map[*resource.Resource]string
+	lock          sync.Mutex
+}
+
+var resourceViews storedViews = storedViews{}
+var allMeters meters = meters{
+	meters:        map[string]*meterExporter{"": &defaultMeter},
+	resourceToKey: map[*resource.Resource]string{nil: ""},
+}
+
+// RegisterResourceView is similar to view.Register(), except that it will
+// register the view across all Resources tracked by the system, rather than
+// simply the default view.
+func RegisterResourceView(views ...*view.View) error {
+	resourceViews.lock.Lock()
+	defer resourceViews.lock.Unlock()
+
+	allMeters.lock.Lock()
+	defer allMeters.lock.Unlock()
+
+	errors := make([]error, 0, len(allMeters.meters))
+
+	for _, meter := range allMeters.meters {
+		if err := meter.m.Register(views...); err != nil {
+			errors = append(errors, err)
+		}
+	}
+	if len(errors) > 0 {
+		return errors[0] // The first error is as good as any
+	}
+
+	resourceViews.views = append(resourceViews.views, views...)
+	return nil
+}
+
+func UnregisterResourceView(views ...*view.View) {
+	resourceViews.lock.Lock()
+	defer resourceViews.lock.Unlock()
+
+	allMeters.lock.Lock()
+	defer allMeters.lock.Unlock()
+
+	for _, meter := range allMeters.meters {
+		meter.m.Unregister(views...)
+	}
+
+	j := 0
+	for _, view := range resourceViews.views {
+		toRemove := false
+		for _, viewToRemove := range views {
+			if view == viewToRemove {
+				toRemove = true
+				break
+			}
+		}
+		if !toRemove {
+			resourceViews.views[j] = view
+			j++
+		}
+	}
+	resourceViews.views = resourceViews.views[:j]
+}
+
+func setFactory(f func(*resource.Resource) (view.Exporter, error)) error {
+	if f == nil {
+		return fmt.Errorf("do not setFactory(nil)!")
+	}
+
+	allMeters.lock.Lock()
+	defer allMeters.lock.Unlock()
+
+	allMeters.factory = f
+
+	errs := make([]error, len(allMeters.meters))
+
+	for r, meter := range allMeters.meters {
+		e, err := f(resourceFromString(r))
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		meter.m.RegisterExporter(e)
+		meter.m.UnregisterExporter(meter.e)
+		meter.e = e
+	}
+	if len(errs) > 0 {
+		return errs[0]
+	}
+	return nil
+}
+
+func flushResourceExporters() {
+	allMeters.lock.Lock()
+	defer allMeters.lock.Unlock()
+
+	for _, meter := range allMeters.meters {
+		flushGivenExporter(meter.e)
+	}
+}
+
+func meterExporterForResource(r *resource.Resource) *meterExporter {
+	key, ok := allMeters.resourceToKey[r]
+	if !ok {
+		key = resourceAsString(r)
+		allMeters.resourceToKey[r] = key
+	}
+
+	mE := allMeters.meters[key]
+	if mE == nil {
+		mE = &meterExporter{}
+		allMeters.meters[key] = mE
+	}
+	if mE.o != nil {
```

Comment:
```suggestion
	} else if mE.o != nil {
```

Reply:
I find it harder to reason about "else if" than independent statements. I suspect that any efficiency gains here are minimal.